edge-client/subdevice_fota.cpp

```diff
@@ -0,0 +1,292 @@
+    if(uri_path) {
+        free(uri_path);
+        uri_path = NULL;
+    }
+}
+
+void get_vendor_id(uint8_t* v_id) {
+    if(fota_ctx) {
+        memcpy(v_id, fota_ctx->fw_info->vendor_id, FOTA_MANIFEST_VENDOR_ID_SIZE);
+    }
+    else {
+        return;
+    }
+}
+void get_class_id(uint8_t* c_id) {
+    if(fota_ctx) {
+        memcpy(c_id, fota_ctx->fw_info->class_id, FOTA_MANIFEST_CLASS_ID_SIZE);
+    }
+    else {
+        return;
+    }
+}
+
+void get_uri(char* c_url) {
+    if(fota_ctx) {
+        memcpy(c_url, fota_ctx->fw_info->uri, FOTA_MANIFEST_URI_SIZE);
+    }
+    else {
+        return;
+    }
+}
+
+
+unsigned int get_component_id() {
+    if(fota_ctx) {
+        return fota_ctx->comp_id;
+    }
+    else {
+        return 0;
+    }
+}
+
+void free_subdev_context_buffers(void)
+{
+    if(fota_ctx->fw_info) {
+        free(fota_ctx->fw_info);
+        fota_ctx->fw_info = NULL;
+    }
+    if(fota_ctx) {
+        free(fota_ctx);
+        fota_ctx = NULL;
+    }
+    if(endpoint){
+        free(endpoint);
+        endpoint = NULL;
+    }
+}
+size_t get_manifest_fw_size() {
+    if(fota_ctx) {
+        return fota_ctx->fw_info->payload_size;
+    }
+    else
+        return 0;
+}
+
+void subdevice_fota_on_manifest(uint8_t* data, size_t data_size, M2MResource* resource) {
+    tr_info("subdevice_fota_on_manifest");
+    const fota_component_desc_t *comp_desc;
+    fota_component_version_t curr_fw_version;
+    uint8_t curr_fw_digest[FOTA_CRYPTO_HASH_SIZE] = {0};
+    int ret = subdevice_init_buff();
+    if (ret) {
+        FOTA_TRACE_DEBUG("Initialising buffer failed");
+        goto fail;
+    }
+    get_endpoint(endpoint, resource->uri_path());
+    ret = fota_manifest_parse(data, data_size,fota_ctx->fw_info);
+    if (ret) {
+        FOTA_TRACE_DEBUG("Pelion FOTA manifest rejected %d", ret);
+        goto fail;
+    }
+
+    FOTA_TRACE_DEBUG("Pelion FOTA manifest is valid");
+
+    ret = fota_component_name_to_id(fota_ctx->fw_info->component_name, &fota_ctx->comp_id);
+    if (ret) {
+        FOTA_TRACE_ERROR("Manifest addresses unknown component %s", fota_ctx->fw_info->component_name);
+        ret = FOTA_STATUS_UNEXPECTED_COMPONENT;
+        goto fail;
+    }
+    
+    fota_component_get_desc(fota_ctx->comp_id, &comp_desc);
+
+    if (comp_desc->desc_info.curr_fw_get_digest) {
+        comp_desc->desc_info.curr_fw_get_digest(curr_fw_digest);
+    }
+
+    fota_component_get_curr_version(fota_ctx->comp_id, &curr_fw_version);
+    FOTA_FI_SAFE_COND(fota_ctx->fw_info->version > curr_fw_version,
+                      FOTA_STATUS_MANIFEST_VERSION_REJECTED, "Manifest payload-version rejected - too old");
+
+    tr_info("Handle manifest: component %s, curr version %" PRIu64 ", new version %" PRIu64 "",
+                     fota_ctx->fw_info->component_name, curr_fw_version, fota_ctx->fw_info->version);
+
+    if (fota_ctx->fw_info->payload_format == FOTA_MANIFEST_PAYLOAD_FORMAT_DELTA) {
+    #if defined(FOTA_DISABLE_DELTA)
```

Comment:
@prakharbhatt007  Any directive should start from the 1st column of the row. See - https://google.github.io/styleguide/cppguide.html#Preprocessor_Directives

Reply:
implemented